### PR TITLE
Fix performance drop when platform is slower to retrieve images w.r.t. requested rate

### PR DESCRIPTION
**Pull request by roncapat**
Whenever `pub_frame_rate` was higher than a certain threshold (5.0 Hz on our Jetson Nano) setting a greater value (like 6.0 or higher) resulted in a sudden drop of framerate from 5 Hz to 1.3 Hz at most.

This patch changes the behaviour that in case of locked mutex skipped the publication of newest frames, skipping a timer tick. Mutex is waited for at most a grab cycle, as the grab cycle is the only other user of that mutex and runs at higher frequency.

### CODE REVIEW

**Comment by Myzhar** [comment]
The PR seems really interesting and promising. Can you please pull the latest modifications to resolve conflicts so I can test the PR before merging?

**Comment by roncapat** [comment]
Yep sorry. I'll close this PR and open another one, but I strongly suggest to do a "squash merge" because we strongly messed up with git and had no time to work it out clean locally.

**Comment by roncapat** [comment]
Now at least the diff is clean and everything is mergeable... let me do a test to verify I didn't messed up the code now, and I'll report to you soon.

**Comment by roncapat** [comment]
@Myzhar now it should build correctly. 

It happens that I can't build this version now because on my eloquent system `image_transport/image_transport.hpp` is still `image_transport/image_transport.h` (and all other headers of that package has .h extension too). If I manually change in the header file those extension, the whole package builds.

**Comment by Myzhar** [comment]
Thank you @roncapat 
I'm going to test this next week and merge it if everything works as expected.
Thank you for your effort

**Comment by roncapat** [comment]
You're welcome! Poll me for any issue

File: zed_components/src/zed_camera/include/zed_camera_component.hpp [hunk 1]
```diff
@@ -584,6 +585,17 @@ class ZedCamera : public rclcpp::Node
...
+    sl::Mat mat_left,mat_left_raw;
+    sl::Mat mat_right,mat_right_raw;
+    sl::Mat mat_left_gray,mat_left_raw_gray;
+    sl::Mat mat_right_gray,mat_right_raw_gray;
+    sl::Mat mat_depth,mat_disp,mat_conf;
+    sl::Timestamp ts_rgb=0;       // used to check RGB/Depth sync
+    sl::Timestamp ts_depth=0;     // used to check RGB/Depth sync
+    sl::Timestamp grab_ts=0;
+    rclcpp::Time timeStamp;
+
```
Comment (Myzhar) [inline zed_components/src/zed_camera/include/zed_camera_component.hpp]: @roncapat why did you move all those declarations in the class definition?

File: zed_components/src/zed_camera/src/zed_camera_component.cpp [hunk 2]
```diff
@@ -3996,23 +3996,23 @@ bool ZedCamera::publishVideoDepth(rclcpp::Time& out_pub_ts)
...
+  static size_t leftGrayRawSubnumber = 0;
+  static size_t rightSubnumber = 0;
+  static size_t rightRawSubnumber = 0;
+  static size_t rightGraySubnumber = 0;
+  static size_t rightGrayRawSubnumber = 0;
+  static size_t stereoSubnumber = 0;
+  static size_t stereoRawSubnumber = 0;
+  static size_t depthSubnumber = 0;
+  static size_t confMapSubnumber = 0;
+  static size_t disparitySubnumber = 0;
```
Comment (Myzhar) [inline zed_components/src/zed_camera/src/zed_camera_component.cpp:4015]: @roncapat why did you add the "static" keyword for all those variables?

File: zed_components/src/zed_camera/src/zed_camera_component.cpp [hunk 3]
```diff
@@ -4043,74 +4043,59 @@ bool ZedCamera::publishVideoDepth(rclcpp::Time& out_pub_ts)
...
+  std::unique_lock<std::timed_mutex> lock(mCamDataMutex, std::defer_lock);
 
-  if (lock.try_lock())
+  if (lock.try_lock_for(std::chrono::milliseconds(1000/mCamGrabFrameRate)))
   {
     if (rgbSubnumber + leftSubnumber + stereoSubnumber > 0)
     {
-      mZed.retrieveImage(mat_left, sl::VIEW::LEFT, sl::MEM::CPU, mMatResolVideo);
-      retrieved = true;
+      retrieved = sl::ERROR_CODE::SUCCESS ==  mZed.retrieveImage(mat_left, sl::VIEW::LEFT, sl::MEM::CPU, mMatResolVideo);
```
Comment (Myzhar) [inline zed_components/src/zed_camera/src/zed_camera_component.cpp:4063]: @roncapat I like this :+1:

File: zed_components/src/zed_camera/src/zed_camera_component.cpp (hunk 2 above)
Comment (roncapat) [reply zed_components/src/zed_camera/src/zed_camera_component.cpp:4015]: They do not need to be zero-initialized at each loop. It is arguable whether they should have static storage or not, in this way they aren't allocated/de-allocated each time on the stack (but since they are POD types this is not a real advantage).
I can revert and just remove zero-initialization which may avoid a bunch of assembly lines from the executable.

File: zed_components/src/zed_camera/include/zed_camera_component.hpp (hunk 1 above)
Comment (roncapat) [reply zed_components/src/zed_camera/include/zed_camera_component.hpp]: Not necessary. I've done that because since they are(?) not POD types it is better to avoid construction/destruction. Maybe those ones are really better at function-scope and static qualifier.

Comment (Myzhar) [reply zed_components/src/zed_camera/include/zed_camera_component.hpp]: What do you mean by "not POD"? I suppose it, but I want to be sure that we are speaking about he same thing

Comment (roncapat) [reply zed_components/src/zed_camera/include/zed_camera_component.hpp]: Plain-Old Data, in this case they are "fundamental types" like size_t so they are just integers.

Comment (roncapat) [reply zed_components/src/zed_camera/include/zed_camera_component.hpp]: Sorry, in the other comment size_t is a POD surely. Those types in contrast are classes with their contructors and destructors, maybe also virtual tables. So allocating/deallocating them each time may have a cost associated.

Comment (Myzhar) [reply zed_components/src/zed_camera/include/zed_camera_component.hpp]: OK, I guessed right. I can understand what you want to achieve. I think it's better to leave them in the function where they are used because that's the only function and `static` is OK

File: zed_components/src/zed_camera/src/zed_camera_component.cpp (hunk 2 above)
Comment (Myzhar) [reply zed_components/src/zed_camera/src/zed_camera_component.cpp:4015]: No, `static` is OK in this case

File: zed_components/src/zed_camera/include/zed_camera_component.hpp (hunk 1 above)
Comment (Myzhar) [reply zed_components/src/zed_camera/include/zed_camera_component.hpp]: Please change it back and add the `static` declaration. Then I will test the PR and merge it

Comment (roncapat) [reply zed_components/src/zed_camera/include/zed_camera_component.hpp]: Agreed. Maybe today I already push the changes.

Comment (roncapat) [reply zed_components/src/zed_camera/include/zed_camera_component.hpp]: Done!

**Review by Myzhar — APPROVED** [review]

**Comment by Myzhar** [comment]
Thank you @roncapat
